plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -360,6 +360,213 @@ struct block_time_tracker {
+         if (fin_policy = _chain->head_pending_finalizer_policy(); fin_policy) {
+            if (!_is_producer_active_finalizer) {
+               _is_producer_active_finalizer = std::ranges::any_of(fin_policy->finalizers, [&](const auto& f) {
+                  return _producers.contains(to_account_name_safe(f.description));
+               });
+            }
+            if (!_other_active_finalizers) {
+               _other_active_finalizers = std::ranges::any_of(fin_policy->finalizers, [&](const auto& f) {
+                  return !_producers.contains(to_account_name_safe(f.description));
+               });
+            }
+         }
+      }
+   }
+
+   // does not validate n is a valid account_name
+   // Used for lookup of producer name, if n is not a valid producer name then the conversion will create an
+   // account_name that doesn't match.
+   static account_name to_account_name_safe(const std::string& n) {
+      // quick conversion without full checks
+      if (n.size() > 12)
+         return {}; // producer names (account_name) are limited to 12 chars, tables and other names can be 13
+      return eosio::chain::string_to_name(n); // n with invalid chars are encoded as 0 without throwing exception
+   }
+
+private:
+   // votes can be signaled before accepted_block_header, accepted_block_header is currently only signaled from the
+   // main thread while votes can be signaled off the main thread. This allows votes to be signaled for a block before
+   // the block is signaled. We could track last accepted block id and correlate that with the vote signal, but simpler
+   // to add a signal_tolerance. Normally the tolerance only needs to be a few milliseconds, but no real harm in making
+   // it larger. Half a block interval is a nice value.
+   static constexpr fc::microseconds                 _signal_tolerance = fc::microseconds{config::block_interval_us / 2};
+   const controller*                                 _chain = nullptr;
+   std::set<chain::account_name>                     _producers;
+   bool                                              _is_producer_active_finalizer = false;
+   // are there any active finalizers that are not part of producers of this node (not expected to be false outside tests)
+   bool                                              _other_active_finalizers = false;
+   alignas(hardware_destructive_interference_sz)
+   std::atomic<bool>                                 _active = false;
+   std::atomic<fc::time_point>                       _last_producer_vote_time;
+   std::atomic<fc::time_point>                       _last_other_vote_time;
+   std::atomic<fc::time_point>                       _first_accepted_block_time_since_last_producer_vote;
+   std::atomic<fc::time_point>                       _first_accepted_block_time_since_last_other_vote;
+   fc::microseconds                                  _production_pause_vote_timeout;
+};
```

Comment:
My review comments offer an alternative structure that I believe is simpler to reason through, but assume an alternative `production_pause_vote_tracker` class which permits an interface to record when blocks and votes and received and query the pause status according to the tracker.

Particularly, the class does not concern itself with matters like what is the definition of a producer vote vs an other vote. And it has no access to the `controller`.

The interface would look like the following:

```cpp
class production_pause_vote_tracker
{
public:
    // By default start with maximum vote_timeout so that production pause mechanism is effectively disabled.
    explicit production_pause_vote_tracker(fc::microseconds block_acceptance_tolerance = fc::milliseconds(250),
                                           fc::microseconds vote_timeout = fc::microseconds::maximum());

    void set_vote_timeout(fc::microseconds vote_timeout);
    void set_block_acceptance_tolerance(fc::microseconds block_acceptance_tolerance);

    // The member functions below can be called concurrently with respect to
    // themselves and one another, but not generally with respect to the functions above.

    bool record_received_other_vote(fc::time_point vote_received_time);
    bool record_received_producer_vote(fc::time_point vote_received_time);
    bool record_received_block(fc::time_point block_received_time);
    void force_unpause();

    struct pause_status
    {
        fc::time_point                latest_other_vote_received_time;
        fc::time_point                latest_producer_vote_received_time;
        std::optional<fc::time_point> earliest_conflicting_block_received_time;

        bool should_pause() const { return earliest_conflicting_block_received_time.has_value(); }
    };

    pause_status check_pause_status(fc::time_point current_time) const;

private:
    // member variables omitted
};
```

I omitted the actual implementation above for clarity. But I have actually put one together already. You can find it here: https://gist.github.com/arhag/cbf7bbe16b1cea43c419cdcb01d23646

Reply:
Thanks. I like that this can be tested independly.